[BUG] ESP32-C3: Fix reboot/crashes on ESP32-C3s when receiving.

crankyoldgit opened the pull request:
- Seems an irrecv gpio of `14` was a poor default for the ESP-C3. Using it causes a reboot.
  - Using `10` for the input GPIO avoids the reboot.
- Ensure when we allocate/use a system timer, that it actually succeeds.
  - ESP32-C3 only has two system timers (0 & 1). Try to automatically cater for that scenario.
  - Add extra debug messages
  - Add a run-time assert to check it succeeded.

Fixes #1751

crankyoldgit committed:
ESP32-C3: Use different default receive pin to avoid reboot.
Seems `14` was a poor default for the ESP-C3. Using it causes a reboot.
Using `10` for the input GPIO avoids the crash.

Fixes #1751

NiKiZe commented:
Maybe use ifdef instead of ifndef to make it easier to read and also add any other special things in the future.

crankyoldgit committed:
Change from ifndef to ifdef

Per review feedback.

crankyoldgit commented:
> Maybe use ifdef instead of ifndef to make it easier to read and also add any other special things in the future.

Would you believe that's what I originally had, then I changed it to `#ifndef` so it left the old/most common default first.

I've changed it to a plain `#ifdef` per suggestion.

Done.

Please take another look.

crankyoldgit committed:
[BUG] Ensure ESP32 system timer is valid before use.
* Add debug messages & a runtime assert to ensure we correctly allocated a system timer in `IRrecv::enableIRIn()`.
* Limit the max number of the system timer to the appropriate size for the ESP32. ESP32-C3's only have two system timers (0-1), the rest have 4 (0-3).
* Done in such a way it should be correct for future ESP32 models.

Ref: https://docs.espressif.com/projects/arduino-esp32/en/latest/api/timer.html

Fixes #1751

crankyoldgit committed:
Handle when `SOC_TIMER_GROUP_TOTAL_TIMERS` is not defined.
Fixes a compile error on older/normal ESP32 Arduino framework/cores.

For #1751

crankyoldgit committed:
Fix linter issue.

crankyoldgit commented:
@NiKiZe Please take another look. This PR has been significantly extended.
